neal/sampler.py

```diff
@@ -148,23 +153,42 @@ def sample(self, bqm, beta_range=None, num_reads=10, sweeps=1000,
                 is chosen.
 
             initial_states (:class:`dimod.SampleSet` or tuple(numpy.ndarray, dict), optional):
-                Initial states provided either as:
+                One or more samples, each defining an initial state for all the
+                problem variables. Initial states are given one per read, but
+                if fewer than `num_reads` initial states are defined, additional
+                values are generated as specified by `initial_states_generator`.
+
+                Initial states are provided either as:
 
                 * :class:`dimod.SampleSet`, or
+
                 * [deprecated] tuple, where the first value is a numpy array of
                   initial states to seed the simulated annealing runs, and the
                   second is a dict defining a linear variable labelling.
                   In tuple format, initial states provided are assumed to use
                   the same vartype the BQM is using.
 
-                If initial states are not provided (set to ``None``), uniform
-                random samples are use. If they are provided, the number of
-                states *must match* the number of requested reads, `num_reads`.
+            initial_states_generator (str, 'none'/'tile'/'random', optional, default='random'):
```

Comment:
since the generator methods are static, why not pass them in directly rather than with strings? That would allow the user to provide their own generators. Also, why not just as functions?

Reply:
Yes, we could support custom states generators *in addition* to these. I didn't go with that immediately for simplicity reasons.

Note that implementation of each of these generators is sampler-specific: for example in Tabu we do not pre-generate the complete initial states sample set, but iterate `num_reads` times, so we can use an actual Python generator. In Neal, if we went with Python generators, we would need to expand them to a NumPy array before sampling, and it's more efficient to do that directly with NumPy, instead of on a per-sample basis.

Reply:
[Here](https://github.com/dwavesystems/dwave-hybrid/blob/master/hybrid/reference/kerberos.py#L75)'s how we did it in Kerberos. Food for thought.